src/misc/unicode.cpp

```diff
@@ -32,416 +32,618 @@
+// Reverse mappings, codepage -> Unicode
+static std::map<uint16_t, code_page_mapping_reverse_t> mappings_reverse_by_codepage = {};
 
 // ***************************************************************************
 // Grapheme type implementation
 // ***************************************************************************
 
 static bool is_combining_mark(const uint32_t code_point)
 {
-    static const std::vector<std::pair<uint16_t, uint16_t>> ranges = {
-        { 0x0300, 0x036f }, // Combining Diacritical Marks
-        { 0x1ab0, 0x1aff }, // Combining Diacritical Marks Extended
-        { 0x1dc0, 0x1dff }, // Combining Diacritical Marks Supplement
-        { 0x20d0, 0x20ff }, // Combining Diacritical Marks for Symbols
-        { 0xfe20, 0xfe2f }, // Combining Half Marks
-    };
-
-    for (const auto &range : ranges)
-        if (code_point >= range.first && code_point <= range.second)
-            return true;
-
-    return false;
+	static const std::vector<std::pair<uint16_t, uint16_t>> ranges = {
+	        {0x0300, 0x036f}, // Combining Diacritical Marks
+	        {0x064b, 0x065f}, // Arabic Combining Marks
+	        {0x1ab0, 0x1aff}, // Combining Diacritical Marks Extended
+	        {0x1dc0, 0x1dff}, // Combining Diacritical Marks Supplement
+	        {0x20d0, 0x20ff}, // Combining Diacritical Marks for Symbols
+	        {0xfe20, 0xfe2f}, // Combining Half Marks
+	};
+
+	for (const auto& range : ranges) {
+		if (code_point >= range.first && code_point <= range.second) {
+			return true;
+		}
+	}
+
+	return false;
 }
 
-Grapheme::Grapheme(const uint16_t code_point) :
-    code_point(code_point),
-    is_empty(false)
+Grapheme::Grapheme(const uint16_t code_point)
+        : code_point(code_point),
+          is_empty(false)
 {
-    // It is not valid to have a combining mark
-    // as a main code point of the grapheme
+	// It is not valid to have a combining mark
+	// as a main code point of the grapheme
 
-    if (is_combining_mark(code_point))
-        Invalidate();
+	if (is_combining_mark(code_point)) {
+		Invalidate();
+	}
 }
 
 bool Grapheme::IsEmpty() const
 {
-    return is_empty;
+	return is_empty;
 }
 
 bool Grapheme::IsValid() const
 {
-    return is_valid;
+	return is_valid;
 }
 
 bool Grapheme::HasMark() const
 {
-    return mark_1st;
+	return !marks.empty();
 }
 
 uint16_t Grapheme::GetCodePoint() const
 {
-    return code_point;
+	return code_point;
+}
+
+void Grapheme::PushInto(std::vector<uint16_t>& str_out) const
+{
+	if (is_empty || !is_valid) {
+		return;
+	}
+
+	str_out.push_back(code_point);
+	for (const auto mark : marks) {
+		str_out.push_back(mark);
+	}
 }
 
 void Grapheme::Invalidate()
 {
-    is_empty = false;
-    is_valid = false;
+	is_empty = false;
+	is_valid = false;
 
-    code_point = unknown_character;
-    mark_1st   = 0;
-    mark_2nd   = 0;
+	code_point = unknown_character;
+	marks.clear();
+	marks_sorted.clear();
 }
 
 void Grapheme::AddMark(const uint16_t code_point)
```

Comment:
The incoming argument `code_point` is ambiguous with `Grapheme` own `code_point` member variable.

Reply:
Changed argument to `in_code_point` - I have seen such a convention in DOSBox Staging, and I think it is sane.

Reply:
Thanks for noting this standard; makes sense and will try to follow as well!